src/host/posix/multicast_routing_manager.cpp

```diff
@@ -0,0 +1,568 @@
+    case OT_BACKBONE_ROUTER_STATE_SECONDARY:
+        Disable();
+        break;
+    case OT_BACKBONE_ROUTER_STATE_PRIMARY:
+        Enable();
+        break;
+    }
+}
+
+void MulticastRoutingManager::HandleBackboneMulticastListenerEvent(otBackboneRouterMulticastListenerEvent aEvent,
+                                                                   const Ip6Address                      &aAddress)
+{
+    switch (aEvent)
+    {
+    case OT_BACKBONE_ROUTER_MULTICAST_LISTENER_ADDED:
+        mMulticastListeners.insert(aAddress);
+        Add(aAddress);
+        break;
+    case OT_BACKBONE_ROUTER_MULTICAST_LISTENER_REMOVED:
+        mMulticastListeners.erase(aAddress);
+        Remove(aAddress);
+        break;
+    }
+}
+
+void MulticastRoutingManager::Update(MainloopContext &aContext)
+{
+    VerifyOrExit(IsEnabled());
+
+    aContext.AddFdToReadSet(mMulticastRouterSock);
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::Process(const MainloopContext &aContext)
+{
+    VerifyOrExit(IsEnabled());
+
+    ExpireMulticastForwardingCache();
+
+    if (FD_ISSET(mMulticastRouterSock, &aContext.mReadFdSet))
+    {
+        ProcessMulticastRouterMessages();
+    }
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::Enable(void)
+{
+    VerifyOrExit(!IsEnabled());
+
+    InitMulticastRouterSock();
+
+    otbrLogResult(OTBR_ERROR_NONE, "%s", __FUNCTION__);
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::Disable(void)
+{
+    mMulticastListeners.clear();
+    FinalizeMulticastRouterSock();
+
+    otbrLogResult(OTBR_ERROR_NONE, "%s", __FUNCTION__);
+}
+
+void MulticastRoutingManager::Add(const Ip6Address &aAddress)
+{
+    VerifyOrExit(IsEnabled());
+
+    UnblockInboundMulticastForwardingCache(aAddress);
+    UpdateMldReport(aAddress, true);
+
+    otbrLogResult(OTBR_ERROR_NONE, "%s: %s", __FUNCTION__, aAddress.ToString().c_str());
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::Remove(const Ip6Address &aAddress)
+{
+    VerifyOrExit(IsEnabled());
+
+    RemoveInboundMulticastForwardingCache(aAddress);
+    UpdateMldReport(aAddress, false);
+
+    otbrLogResult(OTBR_ERROR_NONE, "%s: %s", __FUNCTION__, aAddress.ToString().c_str());
+
+exit:
+    return;
+}
+
+void MulticastRoutingManager::UpdateMldReport(const Ip6Address &aAddress, bool isAdd)
+{
+    struct ipv6_mreq ipv6mr;
+    otbrError        error = OTBR_ERROR_NONE;
+
+    ipv6mr.ipv6mr_interface = mInfraIf.GetIfIndex();
+    aAddress.CopyTo(ipv6mr.ipv6mr_multiaddr);
+    if (setsockopt(mMulticastRouterSock, IPPROTO_IPV6, (isAdd ? IPV6_JOIN_GROUP : IPV6_LEAVE_GROUP), (void *)&ipv6mr,
+                   sizeof(ipv6mr)) != 0)
+    {
+        error = OTBR_ERROR_ERRNO;
+    }
+
+    otbrLogResult(error, "%s: address %s %s", __FUNCTION__, aAddress.ToString().c_str(), (isAdd ? "Added" : "Removed"));
+}
+
+bool MulticastRoutingManager::HasMulticastListener(const Ip6Address &aAddress) const
+{
+    return mMulticastListeners.find(aAddress) != mMulticastListeners.end();
```

Comment:
This is rather a large PR and I'd suggest not to introduce this enhancement in this PR. Better to reuse the existing implementation to ensure the correctness.